### PR TITLE
usb_host keyboard, keys f1-f12, ctrl-left, ctrl-right, ctrl-up, ctrl-down, page down, page up, insert, delete, pause in VT100 CSI 

### PR DESCRIPTION
**Pull request by iyalosovetsky**
I am very impressed with how the USB host keyboard works. For my CNC project, I `ve to added the ability to work with the function keys f1-f12, ctrl-left, ctrl-right, ctrl-up, ctrl-down, page down, page up, insert, delete, pause in the format of command VT 100. Offer my modified code "supervisor/shared/usb/host_keyboard.c"

### CODE REVIEW

**Comment by tannewt** [comment]
Please run pre-commit over your code: https://learn.adafruit.com/improve-your-code-with-pylint/check-your-code-with-pre-commit It should automatically fix these issue before commit.

File: supervisor/shared/usb/host_keyboard.c [hunk 1]
```diff
@@ -90,13 +110,43 @@ STATIC struct keycode_mapper keycode_to_ascii[] = {
     { HID_KEY_ENTER, HID_KEY_SLASH, 0, FLAG_SHIFT, "\n\x1b\177\t _+{}|~:\"~<>?" },
     { HID_KEY_ENTER, HID_KEY_SLASH, 0, 0, "\r\x1b\10\t -=[]\\#;'`,./" },
 
-    { HID_KEY_F1, HID_KEY_F1, 0x1e, 0, }, // help key on xerox 820 kbd
+    // { HID_KEY_F1, HID_KEY_F1, 0x1e, 0, }, // help key on xerox 820 kbd
```
Comment (tannewt) [inline supervisor/shared/usb/host_keyboard.c:113]:
```suggestion
```

File: supervisor/shared/usb/host_keyboard.c [hunk 2]
```diff
@@ -90,13 +110,43 @@ STATIC struct keycode_mapper keycode_to_ascii[] = {
...
 
     { HID_KEY_KEYPAD_DIVIDE, HID_KEY_KEYPAD_DECIMAL, 0, FLAG_NUMLOCK | FLAG_STRING,
       "/\0" "*\0" "-\0" "+\0" "\n\0" CURSOR_END SEP CURSOR_DOWN SEP CURSOR_PGDN SEP CURSOR_LEFT SEP NOTHING SEP CURSOR_RIGHT SEP CURSOR_HOME SEP CURSOR_UP SEP CURSOR_PGDN SEP CURSOR_INS SEP CURSOR_DEL},
     { HID_KEY_KEYPAD_DIVIDE, HID_KEY_KEYPAD_DECIMAL, 0, 0, "/*-+\n1234567890." },
 
-    { HID_KEY_ARROW_RIGHT, HID_KEY_ARROW_UP, 0, FLAG_STRING, CURSOR_RIGHT SEP CURSOR_LEFT SEP CURSOR_DOWN SEP CURSOR_UP },
+    // { HID_KEY_ARROW_RIGHT, HID_KEY_ARROW_UP, 0, FLAG_STRING, CURSOR_RIGHT SEP CURSOR_LEFT SEP CURSOR_DOWN SEP CURSOR_UP },
```
Comment (tannewt) [inline supervisor/shared/usb/host_keyboard.c:119]: Why remove this?

**Review by tannewt — CHANGES_REQUESTED** [review]
Thanks for the improvements! Just a couple suggestions.

**Comment by iyalosovetsky** [comment]
Thank you for the suggestion
1. About HID_KEY_F1. I`ve changed it to going t0 VT100 CSI standards. if
it`s important to leave compatibility with  xerox 820 keyboard I can change
to the previous version.
2. About HID_KEY_ARROW_RIGHT, HID_KEY_ARROW_UP. I spent a couple of hours
getting the arrows to work with ctr and without ctrl, and this option
worked. Perhaps the cause of the error was resolved by the code below. I
can return everything back and test again.
if (!(mapper->flags & FLAG_CTRL) && ctrl) {
continue;
}

Best regards!

ср, 17 янв. 2024 г. в 20:20, Scott Shawcroft ***@***.***>:

**Comment by jepler** [comment]
> . if it`s important to leave compatibility with  xerox 820 keyboard I can change to the previous version.

No, this is not a priority for me.

File: supervisor/shared/usb/host_keyboard.c [hunk 3]
```diff
@@ -78,6 +78,26 @@ struct keycode_mapper {
 #define CURSOR_INS "\e[2~"
 #define CURSOR_DEL "\e[3~"
 
+// https://learn.microsoft.com/ru-ru/windows/console/console-virtual-terminal-sequences
+// https://aperiodic.net/phil/archives/Geekery/term-function-keys/
+#define F1      "\eOP"
+#define F2      "\eOQ"
+#define F3      "\eOR"
+#define F4      "\eOS"
+#define F5      "\e[15~"
```
Comment (jepler) [inline supervisor/shared/usb/host_keyboard.c:87]: Oddly, my documentation ("man 7 urxvt") has "\e[11~" through "\e14~" for F1..F4.  "\eOP" .. "\eOS" are for KP_F1..KP_F4, which do not exist on traditional PC keyboards.  **however** both xterm and xfce4-terminal are behaving like you've implemented here, so I'm sure it's fine.

**Review by jepler — CHANGES_REQUESTED** [review]

**Comment by iyalosovetsky** [comment]
* if (!(mapper->flags & FLAG_CTRL) && ctrl) {
                    continue;
                } *

Without this, CTRL- UP, DOWN, LEFT, RIGHT do not work.
But you are right about

*{ HID_KEY_A, HID_KEY_Z, 'a', 0, NULL} *

*...*


*if (ctrl) {*

* code &= 0x1f;
}*


that may work badly. I need some time to change the code and test it again.

I think the best way is for a..z not to use that construction.



About binary Or (|) you are right, I`ll correct it.

About F1 ... F4 I will need  time to find out more about that problem.


Thank you for your time, good luck!

Sorry for my english


чт, 18 янв. 2024 г. в 19:00, Jeff Epler ***@***.***>:

File: supervisor/shared/usb/host_keyboard.c [hunk 4]
```diff
@@ -243,6 +293,9 @@ STATIC void process_event(uint8_t dev_addr, uint8_t instance, const hid_keyboard
                 if (mapper->flags & FLAG_CTRL && !ctrl) {
                     continue;
                 }
+                if (!(mapper->flags & FLAG_CTRL) && ctrl) {
+                    continue;
+                }
```
[the remark this answers is not in the document]
Comment (iyalosovetsky) [reply supervisor/shared/usb/host_keyboard.c:298]: You`re right!
I`ve changed to

```
                if (!(mapper->flags & FLAG_CTRL) && ctrl && (mapper->flags & FLAG_STRING)) {
                    continue;
                }
```
 and it wors fine.

[the remark this answers is not in the document]
Comment (iyalosovetsky) [reply supervisor/shared/usb/host_keyboard.c:298]: The best way would be to set a range of special keys, but no idea if that's possible or not.

File: supervisor/shared/usb/host_keyboard.c (hunk 3 above)
Comment (iyalosovetsky) [reply supervisor/shared/usb/host_keyboard.c]: done. I've change to "\e[11~" through "\e14~"

File: supervisor/shared/usb/host_keyboard.c [hunk 5]
```diff
@@ -90,13 +110,43 @@ STATIC struct keycode_mapper keycode_to_ascii[] = {
...
+    // { HID_KEY_ARROW_RIGHT, HID_KEY_ARROW_UP, 0, FLAG_STRING, CURSOR_RIGHT SEP CURSOR_LEFT SEP CURSOR_DOWN SEP CURSOR_UP },
+    { HID_KEY_PAUSE, HID_KEY_PAUSE, 0x1a, 0, },
+    { HID_KEY_PAGE_DOWN, HID_KEY_PAGE_DOWN, 0, FLAG_STRING, CURSOR_PGDN },
+    { HID_KEY_PAGE_UP, HID_KEY_PAGE_UP, 0, FLAG_STRING, CURSOR_PGUP },
+    { HID_KEY_HOME, HID_KEY_HOME, 0, FLAG_STRING, CURSOR_HOME },
+    { HID_KEY_END, HID_KEY_END, 0, FLAG_STRING, CURSOR_END },
+    { HID_KEY_INSERT, HID_KEY_INSERT, 0, FLAG_STRING, CURSOR_INS },
+    { HID_KEY_DELETE, HID_KEY_DELETE, 0, FLAG_STRING, CURSOR_DEL },
+
+    { HID_KEY_F1, HID_KEY_F1, 0, FLAG_STRING, F1 },
```
[the remark this answers is not in the document]
Comment (iyalosovetsky) [reply supervisor/shared/usb/host_keyboard.c]: done.

File: supervisor/shared/usb/host_keyboard.c [hunk 6]
```diff
@@ -90,13 +110,43 @@ STATIC struct keycode_mapper keycode_to_ascii[] = {
...
+    { HID_KEY_F6, HID_KEY_F6, 0, FLAG_STRING, F6 },
+    { HID_KEY_F7, HID_KEY_F7, 0, FLAG_STRING, F7 },
+    { HID_KEY_F8, HID_KEY_F8, 0, FLAG_STRING, F8 },
+    { HID_KEY_F9, HID_KEY_F9, 0, FLAG_STRING, F9 },
+    { HID_KEY_F10, HID_KEY_F10, 0, FLAG_STRING, F10 },
+    { HID_KEY_F11, HID_KEY_F11, 0, FLAG_STRING, F11 },
+    { HID_KEY_F12, HID_KEY_F12, 0, FLAG_STRING, F12 },
+    { HID_KEY_PRINT_SCREEN, HID_KEY_PRINT_SCREEN, 0, FLAG_STRING, PRINT_SCREEN },
+
+    { HID_KEY_ARROW_UP, HID_KEY_ARROW_UP, 0 , FLAG_STRING+FLAG_CTRL,CTRL_UP  },
```
[the remark this answers is not in the document]
Comment (iyalosovetsky) [reply supervisor/shared/usb/host_keyboard.c]: done.

**Comment by iyalosovetsky** [comment]
I've tested on my pico by code bellow. Look like good.

```
import microcontroller
#microcontroller.cpu.frequency = 120000000
print(microcontroller.cpu.frequency)
import time
import sys
import board

import usb_host

import supervisor
pp=usb_host.Port(board.GP26, board.GP27)

while True:
  try:  
    proceedCh=''    
    while supervisor.runtime.serial_bytes_available:
        proceedCh += sys.stdin.read(1)
    if proceedCh!='':
      print('proceedCh:',proceedCh,[ord(res) for res in proceedCh])
    time.sleep(0.1)
  except KeyboardInterrupt:
        print('main: will cancel proceedCh:',proceedCh,[ord(res) for res in proceedCh])
        break
       


print("boot end.")
time.sleep(5)

```

**Review by jepler — APPROVED** [review]
Thanks! I fixed a few small items, see the commit messages. Now good to merge, assuming the CI comes up green.

**Comment by iyalosovetsky** [comment]
Thanks!  I have comliled and double checked on my Raspberry pico. All work fine!!

**Pull request merged**